### PR TITLE
Create an extract and test extract script for the pipeline

### PR DESCRIPTION
**Pull request by CharliedeanCS**
Added an extract script and test_extract script to the pipeline folder.

This script;
- Extracts data from the API
- Extracts all plants found on the API,
- Reads and collects only the data we require
- Outputs the collected data into a CSV file

### CODE REVIEW

File: pipeline/extract.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import csv
+
+import requests
+import requests.exceptions
+
+API_URL = "https://data-eng-plants-api.herokuapp.com/plants/"
+
+
+def convert_plant_data_to_csv(plant_list: list[dict]) -> None:
+    """Converts the list of all plant data into one csv file"""
```
Comment (shivani-63) [inline pipeline/extract.py]:
```suggestion
    """Converts the list of all plant data into one csv file."""
```

File: pipeline/extract.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    with open('plant_data.csv', 'w', newline='', encoding="utf-8") as output_file:
+        dict_writer = csv.DictWriter(output_file, plant_list[0].keys())
+        dict_writer.writeheader()
+        dict_writer.writerows(plant_list)
+
+
+def flatten_and_organize_data(plant_dict: dict) -> dict:
+    """Flattens the data and selects only the parts of the data we need"""
```
Comment (shivani-63) [inline pipeline/extract.py]:
```suggestion
    """Flattens the data and selects only the parts of the data we need."""
```

File: pipeline/test_extract.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+             'origin_location': ['-19.32556', '-41.25528', 'Resplendor', 'BR', 'America/Sao_Paulo'],
+             'plant_id': 50, 'recording_taken': '2023-12-18 14:50:56', 'scientific_name':
+             ['Epipremnum aureum'], 'soil_moisture': 97.21896240755082,
+             'temperature': 13.231052699244547}
+
+
+def test_flatten_and_organize_data():
+    """
+    Tests that the flatten and organize dataset functions works as
+    intended and returns a new dict
```
Comment (shivani-63) [inline pipeline/test_extract.py]:
```suggestion
    intended and returns a new dict.
```

File: pipeline/test_extract.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+    output_dict = flatten_and_organize_data(TEST_DATA)
+
+    assert isinstance(output_dict, dict)
+
+    assert output_dict["Botanist Email"] == "carl.linnaeus@lnhm.co.uk"
+
+
+def test_flatten_and_organize_data_empty():
+    """Tests that an error raises if the data is empty"""
```
Comment (shivani-63) [inline pipeline/test_extract.py]:
```suggestion
    """Tests that an error raises if the data is empty."""
```

File: pipeline/extract.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            plant_json = response.json()
+            plant_keys = plant_json.keys()
+
+            if 'error' not in plant_keys:
+                new_plant = flatten_and_organize_data(plant_json)
+                all_plant_data.append(new_plant)
+
+            current_plant += 1
+    except requests.exceptions.JSONDecodeError:
+        while_plants = False
```
Comment (zandersnow14) [inline pipeline/extract.py:68]: Is this accounting for the internal server error we saw for one of the plants? Just wanted to double check before approving!

**Review by shivani-63 — APPROVED** [review]
Looks good Charlie, thank you!

**Pull request merged**